### PR TITLE
add new API: paddle.clone;Tensor.element_size;nn.utils.parameters_to_vector

### PR DESCRIPTION
**Pull request by zhwesky2010**
### PR types

New features
### PR changes

APIs
### Describe

1.New API:  `paddle.clone`
---
Returns a copy of input Tensor. It will always have a Tensor copy. 
    
Tn addition, the OP provides gradient propagation.
```
import paddle

x = paddle.ones([2], stop_gradient=False)
clone_x = paddle.clone(x)

y = clone_x**3
y.backward()
print(clone_x.grad)          # [3]
print(x.grad)                     # [3]
```
   infoflow 2021-12-14 16-01-21

2.New API:  `paddle.Tensor.element_size` 
---
Returns the size in bytes of an element in the Tensor.
```
import paddle

x = paddle.static.data(name='x', shape=[3, 2], dtype='bool')
x.element_size() # 1

x = paddle.static.data(name='x', shape=[3, 2], dtype='int16')
x.element_size() # 2

x = paddle.static.data(name='x', shape=[3, 2], dtype='float16')
x.element_size() # 2

x = paddle.static.data(name='x', shape=[3, 2], dtype='float32')
x.element_size() # 4

x = paddle.static.data(name='x', shape=[3, 2], dtype='float64')
x.element_size() # 8
```

3.New API：`paddle.nn.utils.parameters_to_vector`
---
Flatten parameters to a 1-D Tensor.
```
import paddle
linear = paddle.nn.Linear(10, 15)

paddle.nn.utils.parameters_to_vector(linear.parameters())
# 1-D Tensor: [165]
```

infoflow 2021-12-14 16-02-15

4.New API：`paddle.nn.utils.vector_to_parameters`
---
Transform a Tensor with 1-D shape to the parameters.
```
import paddle
linear1 = paddle.nn.Linear(10, 15, paddle.ParamAttr(paddle.nn.initializer.Constant(3.)))

vec = paddle.nn.utils.parameters_to_vector(linear1.parameters())

linear2 = paddle.nn.Linear(10, 15)
# copy weight of linear1 to linear2
paddle.nn.utils.vector_to_parameters(vec, linear2.parameters())  
# weight: Tensor(shape=[10, 15], dtype=float32, place=CUDAPlace(0), stop_gradient=False,
#                 [[3. , ..., 3. ],
#                  [..., ..., ...],
#                  [3. , ..., 3. ]])
```

infoflow 2021-12-14 16-02-29

### CODE REVIEW

File: python/paddle/tensor/creation.py [hunk 1]
```diff
@@ -1191,6 +1190,35 @@ def assign(x, output=None):
     return tensor.assign(x, output)
 
 
+def clone(x, name=None):
+    """
+    Returns a copy of input Tensor. It will always have a Tensor copy. 
+    
+    Tn addition, the OP provides gradient propagation.
```
Comment (TCChenlong) [inline python/paddle/tensor/creation.py]: 这两句描述不够清楚，可以在改一下

Comment (zhwesky2010) [reply python/paddle/tensor/creation.py]: Done

File: python/paddle/tensor/creation.py [hunk 2]
```diff
@@ -1192,6 +1191,36 @@ def assign(x, output=None):
     return tensor.assign(x, output)
 
 
+def clone(x, name=None):
+    """
+    Returns a copy of input Tensor. It will always have a Tensor copy. 
+    
+    Tn addition, This function is derivable, so gradients will flow back from the output to input.
```
Comment (TCChenlong) [inline python/paddle/tensor/creation.py]: Tn addition -> In addition

Comment (zhwesky2010) [reply python/paddle/tensor/creation.py]: Done, thx

**Review by TCChenlong — DISMISSED** [review]
LGTM

File: python/paddle/nn/utils/transform_parameters.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    """
+    start = 0
+    helper = LayerHelper("vector_to_parameters", **locals())
+    if in_dygraph_mode():
+        with paddle.no_grad():
+            for param in parameters:
+                shape = param.shape
+                numel = reduce(lambda x, y: x * y, shape)
+                end = start + numel
+                slice_data = _C_ops.slice(vec, None, None, 'axes', [0],
```
Comment (Xreki) [inline python/paddle/nn/utils/transform_parameters.py]: 能否用1个`split`代替多个`slice`？

File: python/paddle/nn/utils/transform_parameters.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                                          'infer_flags', [1], 'starts',
+                                          [start], 'ends', [end])
+                _C_ops.reshape2_(slice_data, None, 'shape', shape)
```
Comment (Xreki) [inline python/paddle/nn/utils/transform_parameters.py]: 为什么这里需要显式使用`inplace`接口呢？

File: python/paddle/nn/utils/transform_parameters.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                helper.append_op(
+                    type='assign',
+                    inputs={'X': slice_data},
+                    outputs={'Out': param})
```
Comment (Xreki) [inline python/paddle/nn/utils/transform_parameters.py]: `vector_to_parameters`可能会比较慢

File: python/paddle/nn/utils/transform_parameters.py (hunk 4 above)
Comment (zhouwei25) [reply python/paddle/nn/utils/transform_parameters.py:123]: 对于reshape，inplace和非inplace成本是一样的。inplace是输入、输出一致，
非inplace是有输出，但是输出是share_buffer_with的，所以用两者都一样，这里用的inplace

Comment (zhwesky2010) [reply python/paddle/nn/utils/transform_parameters.py]: `core.ops.reshape2_`是inplace机制，`core.ops.reshape2`是view机制。使用view机制的OP是写死的，目前有4个：
infoflow 2021-12-22 14-54-41

inplace机制下：输入、输出是同一个var。
view机制下：需要创建一个输出var，但输出var会和输入var是share_buffer_with的，所以两者都没有成本。
这里指定用inplace版本的，写法会简单些，不需要创建一个临时输出var

File: python/paddle/nn/utils/transform_parameters.py (hunk 3 above)
Comment (zhwesky2010) [reply python/paddle/nn/utils/transform_parameters.py]: Done，已去掉多个slice

File: python/paddle/nn/utils/transform_parameters.py (hunk 5 above)
Comment (zhwesky2010) [reply python/paddle/nn/utils/transform_parameters.py]: 已去掉assign，改成最简化的写法：

`parameters_to_vector：使用1个concat OP + 完全inplace的reshape_ OP`
`vector_to_parameters：使用1个split OP + 完全inplace的reshape_ OP`

目前全部使用inplace版的reshape_ OP，每个API实质上只有1个`concat或split` OP的耗时，性能可以得到保证了

File: python/paddle/nn/utils/transform_parameters.py (hunk 4 above)
Comment (zhouwei25) [reply python/paddle/nn/utils/transform_parameters.py]: 对于reshape，inplace和view成本是一样的。
inplace机制下：输入、输出是同一个var。
view机制下：虽然创建了一个输出var，但是输出var会和输入是share_buffer_with的，所以用两者成本都几户没有成本，这里我指定了第一种，就不用再去创建一个临时输出，减少代码行数

**Review by lanxianghit — APPROVED** [review]

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]
LGTM for CMakeLists.txt
